Source/DafnyCore/Resolver/PreType/Flows.cs

```diff
@@ -0,0 +1,454 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.IO;
+using JetBrains.Annotations;
+
+namespace Microsoft.Dafny;
+
+record FlowContext(SystemModuleManager SystemModuleManager, ErrorReporter Reporter, bool DebugPrint) {
+  public TextWriter OutputWriter => SystemModuleManager.Options.OutputWriter;
+}
+
+/// <summary>
+/// A "Flow" is a puzzle piece in recomputing types. The "type adjustment" phase defines a set of flows and then
```

Comment:
Sounds like a constraint to me. Consider renaming to `SubsetPhaseConstraint`. Also, the name "type adjustment" doesn't have much meaning to me, considering a name that includes `SubsetType`, like "InferSubsetType"